### PR TITLE
Studies

### PR DESCRIPTION
**Pull request by victor-paumier**
- Create speciality translatable entity
- Create speciality management page (as hobbies)
- Add speciality toggle page by institute
- Add specialities on list (user, matching)
- Add specialities selector on registration & profile (multiple select)
- Add a speciality criteria
- Let manager choose if field is optional or not

### CODE REVIEW

File: src/Form/EventListener/AddStudyFieldSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            FormEvents::PRE_SET_DATA => 'preSetData',
+            FormEvents::PRE_SUBMIT => 'preSubmit',
+        ];
+    }
+
+    public function preSetData(FormEvent $event)
+    {
+        $user = $event->getData();
+
+        $institute = $user
```
Comment (JeremieSamson) [inline src/Form/EventListener/AddStudyFieldSubscriber.php]: IMO you can put the ternary directly in the `addStudyinForm` function

File: src/Form/EventListener/AddStudyFieldSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    public function preSubmit(FormEvent $event)
+    {
+        $data = $event->getData();
+
+        $institute = \array_key_exists('institute', $data)
+            ? $this->instituteRepo->find($data['institute'])
+            : null;
+
+        $this->addStudyInForm($event->getForm(), $institute);
```
Comment (JeremieSamson) [inline src/Form/EventListener/AddStudyFieldSubscriber.php]: same here

File: src/Services/MatchingHelper.php [hunk 3]
```diff
@@ -185,6 +191,27 @@ private function hobbiesScore($hobbiesMentee, $hobbiesMentor)
         return 0;
     }
 
+    /**
+     * @return int|float
+     */
+    private function studiesScore($studiesMentee, $studiesMentor)
```
Comment (JeremieSamson) [inline src/Services/MatchingHelper.php]: typehint and no @return plz

File: src/Services/MatchingHelper.php [hunk 4]
```diff
@@ -185,6 +191,27 @@ private function hobbiesScore($hobbiesMentee, $hobbiesMentor)
...
+    {
+        if (\count($studiesMentee) > 0 && \count($studiesMentor) > 0) {
+            $nbStudiesMenteeMatched = \count(array_intersect($studiesMentee, $studiesMentor));
+            $nbStudiesMentorMatched = \count(array_intersect($studiesMentor, $studiesMentee));
+
+            if ($nbStudiesMenteeMatched == \count($studiesMentee) && $nbStudiesMentorMatched == \count($studiesMentor)) {
```
Comment (JeremieSamson) [inline src/Services/MatchingHelper.php]: Why not checking also type here ?

File: src/Services/MatchingHelper.php [hunk 5]
```diff
@@ -185,6 +191,27 @@ private function hobbiesScore($hobbiesMentee, $hobbiesMentor)
...
+                return 1;
+            } elseif ($nbStudiesMenteeMatched == \count($studiesMentee) || $nbStudiesMentorMatched == \count($studiesMentor)) {
```
Comment (JeremieSamson) [inline src/Services/MatchingHelper.php]: Same question

File: src/Controller/InstituteController.php [hunk 6]
```diff
@@ -333,10 +353,10 @@ public function toggleStudyLink(
 
         if ($institute->getStudies()->contains($study)) {
             $institute->removeStudy($study);
-            $message = $translator->trans('institute.flash.unlink', ['%name%' => $institute->getName()]); //todo : trans
+            $message = $translator->trans('study.flash.unlink', ['%name%' => $institute->getName()]);
```
Comment (JeremieSamson) [inline src/Controller/InstituteController.php]: `$key` and `$translator->trans($key, ['%name%' => $institute->getName()]);`   in return ?

File: src/Services/MatchingHelper.php (hunk 3 above)
Comment (victor-paumier) [reply src/Services/MatchingHelper.php]: ok for typehint.
about @return, no other way to return multiple types

Comment (JeremieSamson) [reply src/Services/MatchingHelper.php]: force return as float with a cast + typehint ?

Comment (victor-paumier) [reply src/Services/MatchingHelper.php]: done

**Pull request merged**